plugins/modules/identity/keycloak/keycloak_user.py

```diff
@@ -0,0 +1,527 @@
+                                             ['token', 'auth_realm', 'auth_username', 'auth_password']]),
+                           required_together=([['auth_realm', 'auth_username', 'auth_password']]))
+
+    result = dict(changed=False, msg='', diff={}, user='')
+
+    # Obtain access token, initialize API
+    try:
+        connection_header = get_token(module.params)
+    except KeycloakError as e:
+        module.fail_json(msg=str(e))
+
+    kc = KeycloakAPI(module, connection_header)
+
+    realm = module.params.get('realm')
+    state = module.params.get('state')
+    username = module.params.get('username')
+    attributes = module.params.get('attributes')
+
+    # attributes in Keycloak have their values returned as lists
+    # via the API. attributes is a dict, so we'll transparently convert
+    # the values to lists.
+    if attributes is not None:
+        for key, val in module.params['attributes'].items():
+            module.params['attributes'][key] = [
+                val] if not isinstance(val, list) else val
+
+    # Filter and map the parameters names that apply to the user
+    user_params = [x for x in module.params
+                   if x not in list(keycloak_argument_spec().keys()) + ['state', 'realm'] and module.params.get(x) is not None]
+
+    found_users = kc.get_users_by_username(username, realm=realm)
+
+    if found_users is None or found_users == []:
+        before_user = {}
+    else:
+        before_user = found_users[0]
+
+    # Build a proposed changeset from parameters given to this module
+    changeset = {}
+
+    for param in user_params:
+        new_param_value = module.params.get(param)
+        old_value = before_user[param] if param in before_user else None
+        if new_param_value != old_value:
+            changeset[camel(param)] = new_param_value
+
+    # Prepare the desired values using the existing values (non-existence results in a dict that is save to use as a basis)
+    desired_user = before_user.copy()
+    desired_user.update(changeset)
+
+    # Cater for when it doesn't exist (an empty dict)
+    if not before_user:
+        if state == 'absent':
+            # Do nothing and exit
+            if module._diff:
+                result['diff'] = dict(before='', after='')
+            result['changed'] = False
+            result['end_state'] = {}
+            result['user'] = result['end_state']
+            result['msg'] = 'User does not exist; doing nothing.'
+            module.exit_json(**result)
+
+        # Process a creation
+        result['changed'] = True
+
+        if module._diff:
+            result['diff'] = dict(before='', after=desired_user)
+
+        if module.check_mode:
+            module.exit_json(**result)
+
+        # create it
+        kc.create_user(desired_user, realm=realm)
+
+        after_user = kc.get_users_by_username(username, realm=realm)[0]
+
+        result['end_state'] = after_user
+        result['user'] = result['end_state']
```

Comment:
Is there a benefit to populating both `'user'` and `'end_state'`?  Can I suggest we just go with `'end_state'`? (as per all the other key cloak modules.

Reply:
AFAIK the `'user'` object is to keep consistency with other module:
```text
    - Deprecated return value, it will be removed in community.general 6.0.0. Please use the return value I(end_state) instead.
```
but some of the module doesn't have such object, so I have no idea for which I should follow or not.
Maybe we can keep consistency with this and remove it on version `6.0.0`?

Reply:
I would rather not introduce it in the first place. It is only left in the other modules because it will be removed eventually, there's no need to add more cruft that needs to be removed :)

Reply:
I agree, this this is a new module (`keycloak_user.py` didn't exist prior to this PR) so I don't think we should be adding it.